### DISCUSSION TITLE
Additional stage of shutdown for automations

### DISCUSSION THREAD

**Discussion by Shulyaka** [home-assistant/architecture#827] [General] [answered]
Hi! 

I would like to implement a 4th stage of shutdown as per this comment: https://github.com/home-assistant/core/issues/41396#issuecomment-704969818
In short, there is a race condition when we run automations at homeassistant_stop event, and various integrations listen to the same event to close their sockets, devices, eventloops, etc. And when we try to use a service from that integration in an automation, we get a failure.

Here is my proposal:
Add additional step after homeassistant_stop event called homeassistant_stop_integrations and change all integrations that waits `homeassistant_stop` to close their services (which is many of them) to listen to this event instead.

Before:
1st stage:
`homeassistant_stop` event
hass.state = CoreState.stopping
hass.is_running = False
hass.is_stopping = True
timeout is 100
Automation trigger on homeassistant stop fires and the integrations begin to shutdown themselves

After:
1st stage:
`homeassistant_stop` event
hass.state = `CoreState.stopping`
hass.is_running = False
hass.is_stopping = True
timeout is 30
Automation trigger on homeassistant stop fires and the integrations do only those tasks that don't disrupt their services

2nd stage:
`homeassistant_stop_integrations` event (new)
hass.state = `CoreState.stopping_integrations` (new)
hass.is_running = False
hass.is_stopping = True
timeout is 100
Integrations begin to do the task for graceful shutdown.

Questions:
1. Should I add the new stage before or after the current `homeassistant_stop` event? This question is essentially more like how do we name the new first and the second stage and which of them would be the original `homeassistant_stop` event.
2. Should hass.is_running be true or false for the new first pre-stop stage? Currently it is true if self.state in (CoreState.starting, CoreState.running).
3. What timeouts should I chose?

**Comment by Shulyaka** [comment]
How about `homeassistant_stop_services` instead of `homeassistant_stop_integrations`? Does anyone have any opinion on this? Can I start working on the pool request?

**Comment by Shulyaka** [comment]
Here is what I suggest: https://github.com/home-assistant/core/pull/83781

**Comment by tetele** [comment]
I was actually working on the same thing, not knowing that there's an initiative already in progress for this problem. My idea for making automations work on HA shutdown was the same: to add another stage that triggers the shutdown event BEFORE stopping all other services (among which probably the blocker is the Recorder).

However...

I would not change the event that all integrations are currently depending on and introduce a breaking change in all of them. Rather, I'd add another event that automations can use for as a trigger (using the same interface).

So an `EVENT_HOMEASSISTANT_STOP_PENDING` or something similar, used only as the automation trigger for the Home Assistant shutdown event defined in automations. Moreover, the proposed PR below would not even trigger any breaking changes, as far as I can tell.

https://github.com/home-assistant/core/pull/91165

**Comment by Shulyaka** [comment]
This is a kind reminder

**Reply by tetele** [reply]
To whom? My PR has been reviewed by 2 people and still not merged. I was told to just wait. And nobody else is getting pings from you to use as a reminder :)

**Reply by Shulyaka** [reply]
To the core team to review the proposal

**Comment by emontnemery** [comment, marked as answer]
@Shulyaka @tetele There are two implementations which both aim to solve this:
https://github.com/home-assistant/core/pull/83781 by @Shulyaka, which is closed
https://github.com/home-assistant/core/pull/91165 by @tetele 

Do you both agree @tetele's PR is the way to go?

If yes, please update the proposal to clearly outline what's implemented by @91165 (if not already done)

**Reply by frenck** [reply]
This is correctly implemented now, and I think does provided the additional requested change.

Going to close this architectural proposal as it has been implemented. If additional comment come out of this, we can reconsider picking up the discussion.

../Frenck
